sdk/core/Azure.Core/src/Shared/TaskExtensions.cs

```diff
@@ -8,12 +8,30 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Runtime.CompilerServices;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace Azure.Core.Pipeline
 {
     internal static class TaskExtensions
     {
+        public static T WaitWithCancellation<T>(this Task<T> task, CancellationToken cancellationToken)
+        {
+            try
+            {
+                task.Wait(cancellationToken);
+            }
+            catch (Exception)
+            {
+                // ignore exception here to rethrow it later
+            }
+
+            return task.GetAwaiter().GetResult();
+        }
+
+        public static WithCancellationAwaitable<T> WithCancellation<T>(this Task<T> task, CancellationToken cancellationToken)
```

Comment:
@pakrym , can you please review this change?

Reply:
What's the benefit of having this method? Avoiding ThrowIfCancellationRequested before the method call?

Reply:
This is for scenario when a single task can be awaited several times, each with its own `CancellationToken`.

Reply:
I still don't understand, in that scenario we would just check every token we need to check explicitly.

This extension method makes it look like the task would be actually cancelled when cancellation token is which is not true and might be confusing to consumers.

Reply:
Good point. I can rename the method into `AwaitWithCancellation` or `ConfigureAwaitWithCancellation`.
Checking isn't enough cause at the moment when `await` is reached it can be that neither task is completed nor cancellation token is canceled.

Reply:
AwaitWithCancellation sounds good.

Reply:
Fixed